helix-view/src/view.rs

```diff
@@ -201,6 +202,53 @@ impl View {
         }
     }
 
+    pub fn offset_coords_to_in_view_for_search(
+        &self,
+        doc: &Document,
+        scrolloff: usize,
+    ) -> Option<(usize, usize)> {
+        let cursor = doc
+            .selection(self.id)
+            .primary()
+            .cursor(doc.text().slice(..));
+
+        let Position { col, row: line } =
+            visual_coords_at_pos(doc.text().slice(..), cursor, doc.tab_width());
+
+        let inner_area = self.inner_area();
+        let last_line = (self.offset.row + inner_area.height as usize).saturating_sub(1);
+        let last_col = self.offset.col + inner_area.width.saturating_sub(1) as usize;
+
+        let new_offset = |scrolloff: usize| {
+            // - 1 so we have at least one gap in the middle.
+            // a height of 6 with padding of 3 on each side will keep shifting the view back and forth
+            // as we type
+            let scrolloff = scrolloff.min(inner_area.height.saturating_sub(1) as usize / 2);
+
+            let row = if line > last_line.saturating_sub(scrolloff) {
+                // scroll down
+                self.offset.row + line - (last_line.saturating_sub(scrolloff))
+            } else if line < self.offset.row + scrolloff {
+                // scroll up
+                line.saturating_sub(scrolloff)
+            } else {
+                self.offset.row
+            };
+
+            let col = if col > last_col.saturating_sub(scrolloff) {
+                // scroll right
+                self.offset.col + col - (last_col.saturating_sub(scrolloff))
+            } else if col < self.offset.col + scrolloff {
+                // scroll left
+                col.saturating_sub(scrolloff)
+            } else {
+                self.offset.col
+            };
+            (row, col)
+        };
```

Comment:
It looks like this code ends up being mostly duplicated between this and the above function. Can we factor out the shared parts into a function they both call?

Reply:
@the-mikedavis 
Hmmm, `offset_coords_to_in_view` is a simplified version of this, more generic function. So I revised `offset_coords_in_view_center`to take an extra argument, and `offset_coords_to_in_view` as a wrapper for it.
Also, I renamed new methods and changed their args.